### ISSUE TITLE
Datastore.observe not work as expected when provided with criteria

### ISSUE THREAD

**Issue by mynameisguy** [aws-amplify/amplify-js#7756]
**Describe the bug**
Running locally only (without syncing the state).
running few observers that should all get triggered by some update.
Observer without criteria work as expected:
`Datastore.observe(model).subsrcibe(msg => console.log(msg))`
Observers with does not:
`Datastore.observe(model, criteriaFunction).subsrcibe(msg => console.log(msg))`

criteriaFunction is identical as used to get the data and works.

**Additional context**
Using amplify in react native - latest version.
checked on android device

**Comment by amhinson** [comment]
The optional 2nd argument for `DataStore.observe` is just meant to be the `id` of an item and not a function ([docs](https://docs.amplify.aws/lib/datastore/real-time/q/platform/js)). What are you trying to achieve with `criteriaFunction`?

**Comment by mynameisguy** [comment]
From looking at the API ref it seems to be supported:
https://aws-amplify.github.io/amplify-js/api/classes/datastore.html#observe

I also looked at the code and if I'm not missing something it looks like it was implemented.

I'm trying to subscribe to changes in only part of the items according to some parameters.

**Comment by amhinson** [comment]
Ah yes, good observation! (pun not intended)

Could you share what you are passing with `criteriaFunction`?

**Comment by mynameisguy** [comment]
😁

Sure.
```
type SomeState @model @auth(rules: [{ allow: owner }]) @key(fields: ["firstKey"]) {
    firstKey: String!
    secKey: String!
    firstVal: Float!
    secVal: Float!
    boolVal: Boolean!
}
```

criteriaFunction:
```
c => c['firstKey']('eq', 'firstKeyValue');
```

changed the names, I'm trying to filter according to state.firstKey === 'firstKeyValue'

**Comment by mynameisguy** [comment]
Tried to reproduce in a new project and encountered a completely opposite issue. Observe get called all the time even when criteria is not correct.

To reproduce, create a new project as described in [docs](https://docs.amplify.aws/lib/datastore/getting-started/q/platform/js), and change to the following:

schema.graphql:
```
type Todo @model {
  id: ID!
  name: String!
  description: String
}
```

App.js:
```
const App: () => React$Node = () => {
  DataStore.observe(Todo).subscribe((msg) =>
    console.log('subscribed activated TODO ALL', msg),
  );
  DataStore.observe(Todo, (c) => c.name === 'todo1').subscribe((msg) =>
    console.log('subscribed activated TODO1', msg),
  );

  const firstTodo1 = new Todo({name: 'todo1', description: 'description1'});
  const firstTodo2 = new Todo({name: 'todo2', description: 'description2'});
  const addTodo1 = () => {
    DataStore.save(firstTodo1);
  };
  const updateTodo1 = () => {
    DataStore.save(
      Todo.copyOf(firstTodo1, (updated) => {
        updated.description = Math.random().toString(16).substr(2, 8);
      }),
    );
  };
  const addTodo2 = () => {
    DataStore.save(new Todo({name: 'todo2', description: 'description2'}));
  };
  const updateTodo2 = () => {
    DataStore.save(
      Todo.copyOf(firstTodo2, (updated) => {
        updated.description = Math.random().toString(16).substr(2, 8);
      }),
    );
  };

  return (
    <>
      <StatusBar barStyle="dark-content" />
      <SafeAreaView>
        <View style={styles.sectionContainer}>
          <Pressable onPress={addTodo1}>
            <Text style={styles.sectionTitle}>addTodo1</Text>
          </Pressable>
          <Pressable onPress={updateTodo1}>
            <Text style={styles.sectionTitle}>updateTodo1</Text>
          </Pressable>
          <Pressable onPress={addTodo2}>
            <Text style={styles.sectionTitle}>addTodo2</Text>
          </Pressable>
          <Pressable onPress={updateTodo2}>
            <Text style={styles.sectionTitle}>updateTodo2</Text>
          </Pressable>
        </View>
      </SafeAreaView>
    </>
  );
};
```

you will notice that when updating/creating new Todo with the name: todo2 both observers get called.

**Comment by amhinson** [comment]
Sorry for the delay here! Thanks for the great reproduction sample. 

With the code that you provided, I do see the observers getting called for every item no matter the name. However, it looks like the 2nd argument to `observe` needs to be a [predicate](https://docs.amplify.aws/lib/datastore/data-access/q/platform/js#predicates) instead of `(c) => c.name === 'todo1'`, i.e.:

```js
DataStore.observe(Todo, (c) => c.name("eq", "todo1"))
```

When I make the above change, the subscription behavior seems to be as expected.

Let me know if you're still having issues 👍

**Comment by mynameisguy** [comment]
Sorry, my bad.

Now it's back to the issue I encountered in the original post.
Adding new Todo with the title "todo1" does activate the subscriber, but updating one does not.

It looks like the criteriaFunc runs on the update message (which only includes the changes) and not the entire object. I'm assuming it's not the expected behavior right?

**Comment by amhinson** [comment]
Hmm that is strange. Are you still using the same sample you provided earlier in the issue, but with the predicate as the 2nd argument to `observe`?

When I run that sample, I am seeing the expected behavior:
- Click "addTodo1" - See events for TODO ALL and TODO1
- Click "updateTodo1" - See events for TODO ALL and TODO1
- Click "addTodo2" - See events _just_ for TODO ALL
- Click "updateTodo1" - See events _just_ for TODO ALL

![Kapture 2021-03-03 at 14 56 10](https://user-images.githubusercontent.com/17091790/109883876-ebb95c00-7c30-11eb-9ed5-dce4e08a6d40.gif)

**Comment by mynameisguy** [comment]
Really strange... I just double checked and I'm still seeing the same behavior:

https://user-images.githubusercontent.com/605165/109938238-786b2500-7cd8-11eb-85ed-2aac09573753.mov

I checked on both android & iOS (the video is of iPhone 12 simulator with iOS 14.4)
updated all the relevant packages, now the package.json:

```
    "dependencies": {
        "@aws-amplify/datastore": "^2.9.11",
        "@react-native-community/netinfo": "^5.9.10",
        "aws-amplify": "^3.3.22",
        "aws-amplify-react-native": "^4.3.1",
        "react": "16.13.1",
        "react-native": "0.63.4"
    },
    "devDependencies": {
        "@babel/core": "^7.8.4",
        "@babel/runtime": "^7.8.4",
        "@react-native-community/eslint-config": "^1.1.0",
        "babel-jest": "^25.1.0",
        "eslint": "^6.5.1",
        "ini": "^1.3.5",
        "inquirer": "^6.5.1",
        "jest": "^25.1.0",
        "metro-react-native-babel-preset": "^0.59.0",
        "react-test-renderer": "16.13.1"
    },
    "jest": {
        "preset": "react-native"
    }
```


the code after changes:
```javascript
const App: () => React$Node = () => {
  DataStore.observe(Todo).subscribe((msg) =>
    console.log('subscribed activated TODO ALL'),
  );
  DataStore.observe(Todo, (c) => c.name('eq', 'todo1')).subscribe((msg) =>
    console.log('subscribed activated TODO1'),
  );

  const firstTodo1 = new Todo({name: 'todo1', description: 'description1'});
  const firstTodo2 = new Todo({name: 'todo2', description: 'description2'});
  const addTodo1 = () => {
    DataStore.save(firstTodo1);
  };
  const updateTodo1 = () => {
    DataStore.save(
      Todo.copyOf(firstTodo1, (updated) => {
        updated.description = Math.random().toString(16).substr(2, 8);
      }),
    );
  };
  const addTodo2 = () => {
    DataStore.save(new Todo({name: 'todo2', description: 'description2'}));
  };
  const updateTodo2 = () => {
    DataStore.save(
      Todo.copyOf(firstTodo2, (updated) => {
        updated.description = Math.random().toString(16).substr(2, 8);
      }),
    );
  };

  return (
    <>
      <StatusBar barStyle="dark-content" />
      <SafeAreaView>
        <View style={styles.sectionContainer}>
          <Pressable onPress={addTodo1}>
            <Text style={styles.sectionTitle}>addTodo1</Text>
          </Pressable>
          <Pressable onPress={updateTodo1}>
            <Text style={styles.sectionTitle}>updateTodo1</Text>
          </Pressable>
          <Pressable onPress={addTodo2}>
            <Text style={styles.sectionTitle}>addTodo2</Text>
          </Pressable>
          <Pressable onPress={updateTodo2}>
            <Text style={styles.sectionTitle}>updateTodo2</Text>
          </Pressable>
        </View>
      </SafeAreaView>
    </>
  );
};
```


BTW, if I change `updateTodo2 ` to:
```javascript
  const updateTodo2 = () => {
    DataStore.save(
      Todo.copyOf(firstTodo2, (updated) => {
        updated.name = 'todo1';
        updated.description = Math.random().toString(16).substr(2, 8);
      }),
    );
  };
```

it triggers `subscribed activated TODO1`

**Comment by amhinson** [comment]
interesting! As a sanity check, could you delete your node_modules and reinstall?

**Comment by mynameisguy** [comment]
Tried just now - still seeing the same behavior...

**Comment by amhinson** [comment]
Would you be able to provide a full repo with your sample, just to confirm we're approaching it in the exact same way?

**Comment by mynameisguy** [comment]
Sure. I also cloned it in a new folder and verified that I still see the issue.
https://github.com/mynameisguy/amplify-datastore-observe-bug

**Comment by amhinson** [comment]
Ah ok, thanks for the repo! I was able to see the behavior you were describing. One interesting key difference though is that the sample you provided didn't actually call `Amplify.configure()`. This means that DataStore is not actually syncing with the backend. However, if I add in `Amplify.configure(config)`, DataStore starting the sync process as expected and I do see the same observable events being triggered in the same way as the video I posted.

Were you intending this app to be offline only, or were you planning on connecting it to your AppSync backend?
